### PR TITLE
Deprecate explicitly setting `metricName` for all scalers

### PR DESCRIPTION
**Pull request by yardenshoham**
In v2.12 this setting will be internalized and should not be set by users

A log message was added

### Checklist

- [X] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [X] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs))
- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Docs PR https://github.com/kedacore/keda-docs/pull/1074
Relates to #4220

### CODE REVIEW

**Review by JorTurFer — CHANGES_REQUESTED** [review]
This is a good idea, but sadly we cannot do it at this level because there are some scalers where `metricName` is a valid value used in the scaler (like aws cloudwatch). We need to do it one by one

**Comment by zroubalik** [comment]
> This is a good idea, but sadly we cannot do it at this level because there are some scalers where `metricName` is a valid value used in the scaler (like aws cloudwatch). We need to do it one by one

If aws cloudwatch is the only one, then we can probably add the exception to the code added in this PR. Or is there any other scaler where metricName is valid?

**Comment by JorTurFer** [comment]
I'm not 100% sure, I haven't had time to check it in depth yet

**Comment by zroubalik** [comment]
@yardenshoham could you please check the source code of scalers that uses `metricName` to verify that it is being used only for internal metrics representation?

For example:
https://github.com/kedacore/keda/blob/72defc212aa5ade7e21303d1dc2a00ab84a2803c/pkg/scalers/prometheus_scaler.go#L48
vs
https://github.com/kedacore/keda/blob/72defc212aa5ade7e21303d1dc2a00ab84a2803c/pkg/scalers/aws_cloudwatch_scaler.go#L36

**Comment by yardenshoham** [comment]
Will do, tonight

**Comment by JorTurFer** [comment]
Also remember that this doesn't solve all the cases. I mean, this adds  the warning notification (which is nice), but scalers like prometheus requires metricName, they need to be updated to make it optional too

**Comment by yardenshoham** [comment]
Yes, I didn't claim this would resolve the issue

**Comment by JorTurFer** [comment]
Yeah, I know it, I said just to share the point. Maybe we can fix them as part of this PR if there are only a few. I mean, if only 1-3 scalers require changes, we could do it in one shoot

**Comment by yardenshoham** [comment]
I think it's a good idea. For prometheus, how about I normalize the query and set that as `queryName`?

**Comment by JorTurFer** [comment]
> I think it's a good idea. For prometheus, how about I normalize the query and set that as `queryName`?

I think it's not necessary, adding a fixed value like 'prometheus' is more than enough as this is an internal value. Something like
If given value, warning and use it
If not given, use something like 'prometheus'

And please, add a comment in scalers where the metricName should be removed or write them somewhere to have them tracked for removing them :pray:

**Comment by yardenshoham** [comment]
So `meta.metricName` need not be unique?

**Comment by JorTurFer** [comment]
> So `meta.metricName` need not be unique?

No, it sin't required since we introduced a dynamic prefix. Each metric has `sX-` as a prefix where X is the scaler index, so even if you use the same name for all the cases, the metric name will be unique in the scaled object because it's based on the scaler index

**Comment by yardenshoham** [comment]
Two scalers stand out:

- aws-cloudwatch
- huawei-cloudeye

How should I handle them?

**Comment by JorTurFer** [comment]
> How should I handle them?

As they are just 2, maybe we can add them in the if to ignore them for the warning. WDYT?

**Comment by yardenshoham** [comment]
Sounds good, done

**Review by JorTurFer — APPROVED** [review]
LGTM! Thanks for the improvement ❤️

**Comment by JorTurFer** [comment]
PTAL @zroubalik

**Comment by yardenshoham** [comment]
> Could you update docs as well?

Sure, I'll get started as soon as this PR lands

**Comment by yardenshoham** [comment]
Should I keep updating this PR?

**Review by zroubalik — APPROVED** [review]
LGTM, great job!

@yardenshoham could you please open a docs PR. We require to have approved docs PR before merging code PR. Thanks a lot

**Comment by zroubalik** [comment]
> Should I keep updating this PR?

It is fine as it now, thanks! :)

**Comment by yardenshoham** [comment]
Great! Docs PR will be ready tomorrow

File: pkg/scalers/couchdb_scaler.go
Comment (zroubalik) [inline pkg/scalers/couchdb_scaler.go:179]: This shouldn't be removed now, just made optional as it is for other scalers.

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: pkg/scalers/couchdb_scaler.go
Comment (zroubalik) [reply pkg/scalers/couchdb_scaler.go:179]: @yardenshoham I've just found this nit^

Comment (yardenshoham) [reply pkg/scalers/couchdb_scaler.go:179]: I removed it because it does nothing as far as I can tell. `meta.metricName` is overidden on the next line

Comment (zroubalik) [reply pkg/scalers/couchdb_scaler.go:179]: you are right, probably some leftover, the parameter is not mentioned in the docs 🤷‍♂️ so we can skip this scaler. Thanks!

**Review by zroubalik — APPROVED** [review]

**Comment by yardenshoham** [comment]
Docs PR https://github.com/kedacore/keda-docs/pull/1074

**Comment by zroubalik** [comment]
> Docs PR [kedacore/keda-docs#1074](https://github.com/kedacore/keda-docs/pull/1074)

awesome, could you please track tasks in the issue #4220 ? In the issue describtion there's TODO list

**Comment by yardenshoham** [comment]
Apart from creating the discussion, what are the other action items? Not sure what needs to be done in "update documentation to specify deprecation notice"

**Comment by zroubalik** [comment]
> Apart from creating the discussion, what are the other action items? Not sure what needs to be done in "update documentation to specify deprecation notice"

exactly this PR you did: https://github.com/kedacore/keda-docs/pull/1074 :)

Just link it in the tasks in the issue and mark this item as done :)

**Comment by yardenshoham** [comment]
I commented https://github.com/kedacore/keda/issues/4220#issuecomment-1443528690

**Comment by zroubalik** [comment]
> I commented [#4220 (comment)](https://github.com/kedacore/keda/issues/4220#issuecomment-1443528690)

I mean the todo list, but I just realized you probably can't modify that, right?

**Comment by yardenshoham** [comment]
Correct

**Comment by yardenshoham** [comment]
Is there any more work to get this PR merged? It has 2 approvals

**Review by zroubalik — APPROVED** [review]
Thanks a lot for the contribution @yardenshoham !

**Pull request merged**
